src/sparse_blas/macros.hpp

```diff
@@ -36,10 +36,91 @@
+        std::int64_t num_rows, std::int64_t num_cols, std::int64_t ld,                \
+        oneapi::mkl::layout dense_layout, FP_TYPE * val);                             \
+    template void set_dense_matrix_data<FP_TYPE>(                                     \
+        sycl::queue & queue, oneapi::mkl::sparse::dense_matrix_handle_t dmhandle,     \
+        std::int64_t num_rows, std::int64_t num_cols, std::int64_t ld,                \
+        oneapi::mkl::layout dense_layout, sycl::buffer<FP_TYPE, 1> val);              \
+    template void set_dense_matrix_data<FP_TYPE>(                                     \
+        sycl::queue & queue, oneapi::mkl::sparse::dense_matrix_handle_t dmhandle,     \
+        std::int64_t num_rows, std::int64_t num_cols, std::int64_t ld,                \
+        oneapi::mkl::layout dense_layout, FP_TYPE * val)
+
+#define INSTANTIATE_COO_MATRIX_FUNCS(FP_TYPE, FP_SUFFIX, INT_TYPE, INT_SUFFIX)                     \
+    template void init_coo_matrix<FP_TYPE, INT_TYPE>(                                              \
+        sycl::queue & queue, oneapi::mkl::sparse::matrix_handle_t * p_smhandle,                    \
+        std::int64_t num_rows, std::int64_t num_cols, std::int64_t nnz,                            \
+        oneapi::mkl::index_base index, sycl::buffer<INT_TYPE, 1> row_ind,                          \
+        sycl::buffer<INT_TYPE, 1> col_ind, sycl::buffer<FP_TYPE, 1> val);                          \
+    template void init_coo_matrix<FP_TYPE, INT_TYPE>(                                              \
+        sycl::queue & queue, oneapi::mkl::sparse::matrix_handle_t * p_smhandle,                    \
+        std::int64_t num_rows, std::int64_t num_cols, std::int64_t nnz,                            \
+        oneapi::mkl::index_base index, INT_TYPE * row_ind, INT_TYPE * col_ind, FP_TYPE * val);     \
+    template void set_coo_matrix_data<FP_TYPE, INT_TYPE>(                                          \
+        sycl::queue & queue, oneapi::mkl::sparse::matrix_handle_t smhandle, std::int64_t num_rows, \
+        std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index,                    \
+        sycl::buffer<INT_TYPE, 1> row_ind, sycl::buffer<INT_TYPE, 1> col_ind,                      \
+        sycl::buffer<FP_TYPE, 1> val);                                                             \
+    template void set_coo_matrix_data<FP_TYPE, INT_TYPE>(                                          \
+        sycl::queue & queue, oneapi::mkl::sparse::matrix_handle_t smhandle, std::int64_t num_rows, \
+        std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index,                    \
+        INT_TYPE * row_ind, INT_TYPE * col_ind, FP_TYPE * val)
+
+#define INSTANTIATE_CSR_MATRIX_FUNCS(FP_TYPE, FP_SUFFIX, INT_TYPE, INT_SUFFIX)                     \
+    template void init_csr_matrix<FP_TYPE, INT_TYPE>(                                              \
```

Comment:
Are you moving them here because they are reused in both Intel oneMKL and cuSPARSE backends? I'm on the edge whether this change is better or if it's better to keep it as it was before, i.e., `#define` these macros, use them immediately, and `#undef` them immediately all in one place.

Reply:
So we used to have other macros used similarly before #500 was merged in recently, for example: https://github.com/oneapi-src/oneMKL/blob/1ce98a699f93bd3a78350269b2e34d822fe43b91/src/sparse_blas/backends/mkl_common/mkl_operations.cxx#L132-L140

It looks like the ones you have here in this PR are now the only ones left, so there are no "other instantiate macros" like I originally thought.

I'm fine with this change, but could you please add a comment above the macro definitions saying that the macros should be undefined immediately after their usage in backend files?

Reply:
Ah, you're right. We can resolve this then.